Remove contract symbol from feed api

holzeis opened the pull request:
Proposes to remove the symbol from the feed api - allowing the consumer to filter for the events interested on the client side. This reduces a lot of complexity in the implementation as well as in the consumption of the API. 

This PR does not yet include the event filtering through a query parameter yet, but should be added in following PRs if we decide to go with that approach.

- [x] Upon connecting to the feed api the latest sent events will be resent, resulting into an issue that either btcusd or ethusd will be sent as the last sent could be either. This results into a delay for btcusd or ethusd to get the latest quote, offer.

Resolves #2673

holzeis commented on daemon-tests/src/flow.rs:
```diff
@@ -39,15 +40,9 @@ pub async fn next_maker_offers(
-        ContractSymbol::EthUsd => rx.eth_usd.clone(),
-    };
-    let maker_offers = next(&mut rx).await?;
+// TODO: what is this function even doing? only called from the tests and the response is ignored.
```
❓ @luckysori @da-kami do you know what this function is used for? It is only used by the tests, but the result is alwas ignored. I would like to remove it, but am not sure if this code is hidding some kind of magic :)

da-kami commented on daemon-tests/src/flow.rs:
```diff
@@ -14,17 +13,19 @@ const NEXT_WAIT_TIME: Duration = Duration::from_secs(if cfg!(debug_assertions) {
 
     let non_empty_offer = |offer: MakerOffers| {
-        if offer.long.is_some() && offer.short.is_some() {
+        if offer.long.is_some()
```
🛠️ I think we should avoid `unwrap`s in the codebase .

Maybe try n `match` for the `Some` case with an additional `if`.
Or maybe an `if let Some(offer) = offer.long` 

---

Is it intentional that we always send out long and short together? (I guess yes)

holzeis replied:
Happy to change, but i figured I can safely unwrap here since I've checked before for is_some()

----

Well I think it is intentional at least that's how it has been.

holzeis replied:
fixed in https://github.com/itchysats/itchysats/pull/2679/commits/8089dac7d577f12f3bb2688b9fefc586a4b64e27

da-kami commented on daemon-tests/src/flow.rs:
```diff
@@ -39,15 +40,9 @@ pub async fn next_maker_offers(
-        ContractSymbol::EthUsd => rx.eth_usd.clone(),
-    };
-    let maker_offers = next(&mut rx).await?;
+// TODO: what is this function even doing? only called from the tests and the response is ignored.
```
This function ensures that (initially) the value consumed from the feed is `None`.
This is used for sanity in the tests. I think we should keep it, but am open to discussing better solutions.

holzeis replied:
well as mentioned the response if true or false is no where used. So if we really want to ensure that we should add an assert.

luckysori replied:
I agree that it wasn't working as intended but I think what we should do is simply bubble up an error if the condition is not met.

That means the return type would be `Result<()>` and the caller can just unwrap since we're in a test, and ignore the `Ok(())`.

I would vote against `assert`ing because that should really only be used to verify whatever the test is actually testing, rather than to check if there's  a bug in the implementation of the test.

da-kami reviewed (commented):
Thanks, great work! I think it's definitely better to let the caller decide and identify the sent data both the symbol.

I'll have another look once the unwrap is gone.

luckysori commented on maker-frontend/src/MakerApp.tsx:
```diff
@@ -63,14 +63,23 @@ export default function App() {
+    const makerLongOrder = useLatestEvent<MakerOffer>(
+        source,
+        "long_offer",
+        undefined,
```
❓ Why this `undefined` here but not below?

luckysori replied:
It looks like you were meaning to use the `filter` optional argument below, so the `undefined` is missing there.

holzeis replied:
good catch, I've forgot the undefined below.

luckysori commented on maker-frontend/src/components/Hooks.tsx:
```diff
@@ -5,6 +5,7 @@ export default function useLatestEvent<T,>(
     source: EventSource,
     event_name: string,
     mapping: (key: string, value: any) => any = (key, value) => value,
+    filter?: (event: Event) => boolean,
```
❓ What if we don't introduce this filter and rely on the caller to pass in a source with already filtered events?

Alternatively we could also move the filter to the `event_name` itself, with names such as `btc_usd_short_offer`.

holzeis replied:
I haven't found a way to filter events on the event source - happy to change if you know how. Regarding the alternative, I am not so sure about it. I can think of argument for and against it. I guess my key concern here is that this requires our backend to make assumptions about how events are consumed by the client. It might bring benefits for our case, but it might be cumbersome for others. Our events are currently simple changed events - adding what has changed into the event could get very complex very quickly (because where do you draw the line?) Instead I'd suggest we introduce server side filtering using e.g. a query parameter that could also resolve the need for client side filtering.

luckysori commented on daemon/src/taker_cfd.rs:
```diff
@@ -93,19 +92,11 @@ impl Actor {
+            .send(projection::Update(
                 takers_perspective_of_maker_offers.clone(),
-            )))
+            ))
```
👍 I think these simplifications of `taker_cfd` and `projection` are in line with what I've done in https://github.com/itchysats/itchysats/pull/2680. 

We'll have to adapt some code to merge them together, but I'm now sending lists of pure `model::Offer`s around, which is a step further than what you've done here.

luckysori commented on maker-frontend/src/MakerApp.tsx:
```diff
@@ -63,14 +63,23 @@ export default function App() {
+    const makerShortOrder = useLatestEvent<MakerOffer>(
+        source,
+        "short_offer",
+        (data: any) => data && data.contract_symbol == "BtcUsd",
```
❌ If we hard-code this to `BtcUsd` we will not filter the correct offers when we flip the switch. I think you need to use the variable `symbol` on line 63.

holzeis replied:
It was hardcoded before so I did not change that as it wasn't part of the issue. Happy to change though.

luckysori replied:
AFAIK it wasn't hard-coded: we had a dynamic source based on the value of `symbol`, which changed every time `setSymbol` was called.

The way I see it, with this changes when we change the symbol to `ETHUSD` we wills till display `BTCUSD`.

holzeis replied:
you are right.. must have changed it myself and thought it has been like that before 🙈

luckysori commented on daemon-tests/src/flow.rs:
```diff
@@ -21,15 +21,25 @@ pub async fn next_maker_offers(
+            None => return None,
+        }
+
+        Some(offer)
```
🙃 It's a nice idea to use pattern matching here, but you can even go a bit further. Let me show you something that I consider more Rust-idiomatic:

```rust
let non_empty_offers = |offers: MakerOffers| match &offers {
    MakerOffers {
        long: Some(long),
        short: Some(short),
    } if &long.contract_symbol == contract_symbol
        && &short.contract_symbol == contract_symbol =>
    {
        Some(offers)
    }

    _ => None,
};
```

There are a couple things going on here that might be interesting:

- Pattern matching against a custom struct, which can be really powerful.
- Match [guards](https://doc.rust-lang.org/rust-by-example/flow_control/match/guard.html).
- Avoiding explicit returns. I think idiomatic Rust uses returns sparingly. I'd say this team strives to use them mostly in early returns or not at all. The reason being that it's easier to read a function when you can expect it to return from only one or two predictable places.

holzeis replied:
thank you for that rust lesson 😃

luckysori reviewed (commented):
Looks good! 

I think there's one thing that seems wrong. The rest is less important.

luckysori reviewed (approved):
LGTM

da-kami commented on taker/src/routes.rs:
```diff
@@ -57,37 +56,9 @@ pub struct IdentityInfo {
 }
 
-#[derive(Debug, Copy, Clone, strum_macros::Display)]
-pub enum ContractSymbol {
```
🙃 I think having a dedicated struct for the API here is actually a good thing, because we can decide the serialization can be guaranteed in a better way. (e.g. not breaking the API with renaming a field)

luckysori replied:
I agree on principle, but this was removed because it (or anything like it) is not used at all after these changes.

da-kami replied:
My concern was that we expose `model::ContractSymbol` as part of `projection::CfdOffers`, but that was already the case before.
Please ignore this for now.

da-kami commented on daemon/src/projection.rs:
```diff
@@ -108,43 +98,26 @@ impl Actor {
-        });
-        let (tx_quote_btc_usd, rx_quote_btc_usd) = watch::channel(None);
-        let (tx_order_eth_usd, rx_order_eth_usd) = watch::channel(MakerOffers {
+        let (tx_order, rx_order) = watch::channel(MakerOffers {
```
Please rename to `offer` here, feel free to do in follow up

luckysori committed:
feat(feed): Remove contract symbol from offer feed path

It should make it easier for us to implement and consume this feed if
the client simply filters the events they're interested in.

The filtering could eventually be achieved through a query parameter
on the feed URL, but currently we just filter on the client side.

This change temporarily introduces a regression: If a client
reconnects to the feed (e.g. by refreshing the UI) they will only get
the last BTCUSD or ETHUSD offer.

Co-authored-by: Richard Holzeis <richard@coblox.tech>
